improve performance

Uzlopak opened the pull request:
@vkarpov15 

To improve the performance of mongoose, I optimized mpath.get()

Benchmarks:

Before:
```
mpath.get("first", obj) x 7,258,829 ops/sec ±1.02% (89 runs sampled)
mpath.get("first.second", obj) x 3,276,021 ops/sec ±0.54% (91 runs sampled)
mpath.get("first.second.third.1.name", obj) x 1,597,895 ops/sec ±0.53% (92 runs sampled)
mpath.get("comments", obj) x 4,858,920 ops/sec ±0.64% (90 runs sampled)
mpath.get("comments.1", obj) x 3,531,918 ops/sec ±0.55% (89 runs sampled)
mpath.get("comments.2.name", obj) x 2,494,564 ops/sec ±0.67% (91 runs sampled)
mpath.get("comments.2.comments.1.comments.0.val", obj) x 1,132,979 ops/sec ±1.30% (93 runs sampled)
mpath.get("comments.name", obj) x 1,320,459 ops/sec ±0.87% (90 runs sampled)
```

After:
```
mpath.get("first", obj) x 59,995,806 ops/sec ±0.47% (95 runs sampled)
mpath.get("first.second", obj) x 4,726,077 ops/sec ±0.49% (96 runs sampled)
mpath.get("first.second.third.1.name", obj) x 3,373,241 ops/sec ±0.40% (92 runs sampled)
mpath.get("comments", obj) x 23,057,858 ops/sec ±1.34% (87 runs sampled)
mpath.get("comments.1", obj) x 4,074,548 ops/sec ±0.31% (93 runs sampled)
mpath.get("comments.2.name", obj) x 3,708,106 ops/sec ±0.44% (89 runs sampled)
mpath.get("comments.2.comments.1.comments.0.val", obj) x 2,740,973 ops/sec ±0.45% (94 runs sampled)
mpath.get("comments.name", obj) x 1,609,356 ops/sec ±0.23% (95 runs sampled)
```

I would add more optimizations if this one gets accepted. ;)

Uzlopak commented:
@AbdelrahmanHafez
:)

Uzlopak commented:
I really wonder why retrieving "comments" is half the speed of "first". Maybe a deoptimization?!

AbdelrahmanHafez commented on lib/stringToParts.js:
```diff
@@ -1,46 +1,59 @@
     // square brackets contains something other than a number.
-    if (state === 'IN_SQUARE_BRACKETS' && !/\d/.test(str[i]) && str[i] !== ']') {
-      state = 'DEFAULT';
+    if (state === 1 && char !== ']' && !/\d/.test(char)) {
```
@Uzlopak Should we add a `numericalRegEx`  here instead of creating a new regex each iteration of the loop?

Uzlopak replied:
Benchmarks did not show any perf impact, I think. But sure will modify and bench again.

AbdelrahmanHafez reviewed (approved):
Great work, thanks. 👍

Uzlopak commented:
Added benchmarks for stringToParts

Before:

```
stringToParts("first") x 11,072,720 ops/sec ±1.06% (86 runs sampled)
stringToParts("first.second") x 5,598,882 ops/sec ±0.31% (92 runs sampled)
stringToParts("first.second.third.1.name") x 3,068,168 ops/sec ±0.52% (91 runs sampled)
stringToParts("comments") x 7,866,749 ops/sec ±0.59% (93 runs sampled)
stringToParts("comments.1") x 7,193,658 ops/sec ±0.44% (95 runs sampled)
stringToParts("comments.2.name") x 5,328,157 ops/sec ±0.63% (91 runs sampled)
stringToParts("comments.2.comments.1.comments.0.val") x 2,332,454 ops/sec ±0.58% (91 runs sampled)
stringToParts("comments.name") x 5,605,289 ops/sec ±0.47% (91 runs sampled)
stringToParts("foo[1mystring]") x 2,805,068 ops/sec ±0.52% (90 runs sampled)
stringToParts("foo[1mystring].bar[1]") x 1,854,413 ops/sec ±0.48% (94 runs sampled)
stringToParts("foo[1mystring][2]") x 2,034,563 ops/sec ±0.53% (90 runs sampled)
```

After

```
stringToParts("first") x 72,483,520 ops/sec ±1.16% (93 runs sampled)
stringToParts("first.second") x 6,557,565 ops/sec ±0.46% (89 runs sampled)
stringToParts("first.second.third.1.name") x 6,602,806 ops/sec ±0.38% (94 runs sampled)
stringToParts("comments") x 71,338,213 ops/sec ±1.23% (91 runs sampled)
stringToParts("comments.1") x 6,600,806 ops/sec ±0.37% (95 runs sampled)
stringToParts("comments.2.name") x 6,694,227 ops/sec ±0.46% (92 runs sampled)
stringToParts("comments.2.comments.1.comments.0.val") x 6,632,548 ops/sec ±0.40% (94 runs sampled)
stringToParts("comments.name") x 6,595,215 ops/sec ±0.38% (92 runs sampled)
stringToParts("foo[1mystring]") x 3,924,286 ops/sec ±1.03% (89 runs sampled)
stringToParts("foo[1mystring].bar[1]") x 2,894,064 ops/sec ±0.92% (89 runs sampled)
stringToParts("foo[1mystring][2]") x 3,163,046 ops/sec ±1.01% (88 runs sampled)
```

As you can see, the performance for splitting is slighty slower, but more stable when we have more dots.

Uzlopak commented:
I use now a regex as delimiter:

```
stringToParts("first") x 72,895,088 ops/sec ±1.15% (94 runs sampled)
stringToParts("first.second") x 9,295,418 ops/sec ±0.50% (91 runs sampled)
stringToParts("first.second.third.1.name") x 5,661,157 ops/sec ±0.45% (95 runs sampled)
stringToParts("comments") x 71,932,231 ops/sec ±0.99% (92 runs sampled)
stringToParts("comments.1") x 9,847,519 ops/sec ±0.89% (93 runs sampled)
stringToParts("comments.2.name") x 7,918,027 ops/sec ±0.43% (94 runs sampled)
stringToParts("comments.2.comments.1.comments.0.val") x 4,387,123 ops/sec ±0.42% (93 runs sampled)
stringToParts("comments.name") x 9,477,638 ops/sec ±0.32% (94 runs sampled)
stringToParts("foo[1mystring]") x 4,491,337 ops/sec ±0.90% (88 runs sampled)
stringToParts("foo[1mystring].bar[1]") x 3,401,067 ops/sec ±1.31% (88 runs sampled)
stringToParts("foo[1mystring][2]") x 3,747,096 ops/sec ±1.21% (91 runs sampled)
```

So the performance degrades with every additional dot. 
If we use a dot as string for a delimiter, there is no performance degredation. 

@vkarpov15 What is the better tradeoff?

Uzlopak commented:
I reverted that change. I think it is better to have a stable  performance, than have a very fast degrading performance. But it is your call.

Uzlopak commented:
benchmark for set had wrong parameter order

benchmark of set against master
```
mpath.set("first", obj, val) x 4,888,609 ops/sec ±0.94% (92 runs sampled)
mpath.set("first.second", obj, val) x 2,833,099 ops/sec ±0.31% (96 runs sampled)
mpath.set("first.second.third.1.name", obj, val) x 1,307,018 ops/sec ±0.20% (93 runs sampled)
mpath.set("comments", obj, val) x 3,149,258 ops/sec ±0.37% (92 runs sampled)
mpath.set("comments.1", obj, val) x 3,079,246 ops/sec ±0.25% (94 runs sampled)
mpath.set("comments.2.name", obj, val) x 2,208,094 ops/sec ±0.23% (98 runs sampled)
mpath.set("comments.2.comments.1.comments.0.val", obj, val) x 1,051,480 ops/sec ±0.36% (93 runs sampled)
mpath.set("comments.name", obj, val) x 2,047,667 ops/sec ±0.21% (96 runs sampled)
```

Uzlopak commented:
Improved set

Benchmarks after

```
mpath.set("first", obj, val) x 25,653,711 ops/sec ±0.41% (95 runs sampled)
mpath.set("first.second", obj, val) x 4,050,477 ops/sec ±0.47% (92 runs sampled)
mpath.set("first.second.third.1.name", obj, val) x 2,673,824 ops/sec ±0.32% (94 runs sampled)
mpath.set("comments", obj, val) x 14,821,293 ops/sec ±0.53% (93 runs sampled)
mpath.set("comments.1", obj, val) x 3,316,825 ops/sec ±0.49% (91 runs sampled)
mpath.set("comments.2.name", obj, val) x 3,012,609 ops/sec ±0.33% (92 runs sampled)
mpath.set("comments.2.comments.1.comments.0.val", obj, val) x 2,145,037 ops/sec ±0.43% (92 runs sampled)
mpath.set("comments.name", obj, val) x 3,023,309 ops/sec ±0.25% (93 runs sampled)
```

Uzlopak commented:
@vkarpov15 
@AbdelrahmanHafez 

Ok, I am done. Everything is imho up to standard. :). Can you review it please?

AbdelrahmanHafez commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,21 @@
+    strategy:
+      matrix:
+        os: [ubuntu-latest]
+        node: ['12', '14', '16']
```
We need to add node versions down to v4 to guarantee that we're supporting those versions to remain backward-compatible.

Otherwise, some changes might be using a syntax that's not supported in older versions and will break apps out there.

Uzlopak replied:
I increased in package.json the necessary Version to 12. 

To be Backwards compatible we need to change the lets and consts to var and use the old mocha version. It has perf impacts to use vars. 

But like I said... It is an easy change. 

So do we really need to support from 4 and above? Or can we drop support for non lts versions of node?

AbdelrahmanHafez replied:
We _could_ drop the support of the older versions, but we'd need to release a new major version to do that.

This would be fine for mongoose since we only support Node >=12, but if another package out there is relying on _mpath_, that means they won't be able to support any version lower than node 12.

A solution that gets us the best of both worlds would be:
1- Release a minor version with backward-compatible performance improvements (_var_ instead of _let/const_).
2- Release a major version with the least support of Node 12* where we can use _let/const_.

\* Instead of picking an arbitrary version (12), we should support as many versions as we can.
So, for example, if all the optimizations work with node 6, there would be no reason for us to declare version 6 unsupported. 

Using an older mocha should be fine, are there any new features in mocha that we're relying on?

Uzlopak replied:
I agree.

Uzlopak replied:
Done... it is now working in node 4 like it is working in node 16

Uzlopak commented:
on node 14
master
```
stringToParts("first") x 11,226,719 ops/sec ±0.59% (89 runs sampled)
stringToParts("first.second") x 5,564,152 ops/sec ±0.39% (89 runs sampled)
stringToParts("first.second.third.1.name") x 3,071,618 ops/sec ±0.56% (86 runs sampled)
stringToParts("comments") x 7,853,022 ops/sec ±0.43% (93 runs sampled)
stringToParts("comments.1") x 7,231,409 ops/sec ±0.39% (93 runs sampled)
stringToParts("comments.2.name") x 5,166,704 ops/sec ±0.43% (94 runs sampled)
stringToParts("comments.2.comments.1.comments.0.val") x 2,311,376 ops/sec ±0.50% (94 runs sampled)
stringToParts("comments.name") x 5,772,828 ops/sec ±0.52% (95 runs sampled)
stringToParts("foo[1mystring]") x 2,826,174 ops/sec ±0.63% (90 runs sampled)
stringToParts("foo[1mystring].bar[1]") x 1,890,382 ops/sec ±0.37% (94 runs sampled)
stringToParts("foo[1mystring][2]") x 2,109,377 ops/sec ±0.56% (92 runs sampled)
mpath.get("first", obj) x 6,475,577 ops/sec ±0.83% (89 runs sampled)
mpath.get("first.second", obj) x 2,798,907 ops/sec ±0.51% (92 runs sampled)
mpath.get("first.second.third.1.name", obj) x 1,359,754 ops/sec ±0.53% (91 runs sampled)
mpath.get("comments", obj) x 4,047,669 ops/sec ±0.77% (93 runs sampled)
mpath.get("comments.1", obj) x 2,953,036 ops/sec ±0.78% (91 runs sampled)
mpath.get("comments.2.name", obj) x 2,161,514 ops/sec ±0.39% (90 runs sampled)
mpath.get("comments.2.comments.1.comments.0.val", obj) x 973,012 ops/sec ±0.68% (90 runs sampled)
mpath.get("comments.name", obj) x 1,182,390 ops/sec ±0.50% (95 runs sampled)
mpath.has("first", obj) x 6,320,144 ops/sec ±0.97% (91 runs sampled)
mpath.has("first.second", obj) x 1,995,017 ops/sec ±0.48% (91 runs sampled)
mpath.has("first.second.third.1.name", obj) x 1,064,239 ops/sec ±0.38% (97 runs sampled)
mpath.has("comments", obj) x 3,071,560 ops/sec ±0.60% (94 runs sampled)
mpath.has("comments.1", obj) x 2,725,564 ops/sec ±0.43% (96 runs sampled)
mpath.has("comments.2.name", obj) x 1,833,013 ops/sec ±0.43% (97 runs sampled)
mpath.has("comments.2.comments.1.comments.0.val", obj) x 865,523 ops/sec ±0.58% (95 runs sampled)
mpath.has("comments.name", obj) x 1,949,026 ops/sec ±0.43% (95 runs sampled)
mpath.has("foo[1mystring]", obj) x 1,485,573 ops/sec ±0.52% (96 runs sampled)
mpath.has("foo[1mystring].bar[1]", obj) x 1,231,943 ops/sec ±0.45% (95 runs sampled)
mpath.has("foo[1mystring][2]", obj) x 1,284,794 ops/sec ±0.47% (93 runs sampled)
mpath.has(["comments", "2", "comments","1","comments","0","val"], obj) x 8,859,907 ops/sec ±0.56% (92 runs sampled)
mpath.set("first", obj, val) x 4,373,948 ops/sec ±0.41% (94 runs sampled)
mpath.set("first.second", obj, val) x 2,543,945 ops/sec ±0.27% (93 runs sampled)
mpath.set("first.second.third.1.name", obj, val) x 1,317,064 ops/sec ±0.52% (92 runs sampled)
mpath.set("comments", obj, val) x 3,572,302 ops/sec ±0.30% (96 runs sampled)
mpath.set("comments.1", obj, val) x 2,724,744 ops/sec ±0.22% (91 runs sampled)
mpath.set("comments.2.name", obj, val) x 1,976,067 ops/sec ±0.50% (92 runs sampled)
mpath.set("comments.2.comments.1.comments.0.val", obj, val) x 970,860 ops/sec ±0.29% (94 runs sampled)
mpath.set("comments.name", obj, val) x 1,856,040 ops/sec ±0.18% (95 runs sampled)
mpath.set("arr.test", [[42]], obj)) x 2,486,191 ops/sec ±0.41% (94 runs sampled)
To get some reliable benchmarks, replace in unset return delete cur[part]; with return true.
mpath.unset("first", obj) x 3,923,270 ops/sec ±0.44% (91 runs sampled)
mpath.unset("first.second", obj) x 2,772,676 ops/sec ±0.46% (94 runs sampled)
mpath.unset("first.second.third.1.name", obj) x 1,860,037 ops/sec ±0.50% (95 runs sampled)
mpath.unset("comments", obj) x 3,113,825 ops/sec ±0.51% (92 runs sampled)
mpath.unset("comments.1", obj) x 2,929,182 ops/sec ±0.42% (94 runs sampled)
mpath.unset("comments.2.name", obj) x 2,493,021 ops/sec ±0.54% (91 runs sampled)
mpath.unset("comments.2.comments.1.comments.0.val", obj) x 1,435,760 ops/sec ±0.66% (91 runs sampled)
mpath.unset("comments.name", obj) x 2,574,483 ops/sec ±0.33% (94 runs sampled)
```
this branch
```
> node ./benchmarks

stringToParts("first") x 73,568,537 ops/sec ±0.25% (93 runs sampled)
stringToParts("first.second") x 6,592,612 ops/sec ±0.30% (95 runs sampled)
stringToParts("first.second.third.1.name") x 6,599,756 ops/sec ±0.33% (93 runs sampled)
stringToParts("comments") x 72,313,409 ops/sec ±0.55% (92 runs sampled)
stringToParts("comments.1") x 6,609,289 ops/sec ±0.28% (94 runs sampled)
stringToParts("comments.2.name") x 6,602,317 ops/sec ±0.28% (94 runs sampled)
stringToParts("comments.2.comments.1.comments.0.val") x 6,386,758 ops/sec ±0.35% (95 runs sampled)
stringToParts("comments.name") x 6,622,182 ops/sec ±0.26% (93 runs sampled)
stringToParts("foo[1mystring]") x 4,565,073 ops/sec ±0.53% (91 runs sampled)
stringToParts("foo[1mystring].bar[1]") x 3,441,588 ops/sec ±0.55% (89 runs sampled)
stringToParts("foo[1mystring][2]") x 3,991,700 ops/sec ±0.61% (92 runs sampled)
mpath.get("first", obj) x 26,399,064 ops/sec ±1.23% (83 runs sampled)
mpath.get("first.second", obj) x 4,424,826 ops/sec ±0.30% (93 runs sampled)
mpath.get("first.second.third.1.name", obj) x 3,396,147 ops/sec ±0.39% (93 runs sampled)
mpath.get("comments", obj) x 21,574,192 ops/sec ±0.89% (89 runs sampled)
mpath.get("comments.1", obj) x 3,909,708 ops/sec ±0.35% (92 runs sampled)
mpath.get("comments.2.name", obj) x 3,681,866 ops/sec ±0.42% (92 runs sampled)
mpath.get("comments.2.comments.1.comments.0.val", obj) x 2,824,078 ops/sec ±0.59% (94 runs sampled)
mpath.get("comments.name", obj) x 1,527,945 ops/sec ±0.97% (91 runs sampled)
mpath.has("first", obj) x 34,901,892 ops/sec ±4.62% (69 runs sampled)
mpath.has("first.second", obj) x 4,375,331 ops/sec ±0.34% (91 runs sampled)
mpath.has("first.second.third.1.name", obj) x 3,422,446 ops/sec ±0.51% (95 runs sampled)
mpath.has("comments", obj) x 18,468,901 ops/sec ±1.29% (84 runs sampled)
mpath.has("comments.1", obj) x 4,181,470 ops/sec ±0.40% (94 runs sampled)
mpath.has("comments.2.name", obj) x 3,859,458 ops/sec ±0.45% (91 runs sampled)
mpath.has("comments.2.comments.1.comments.0.val", obj) x 3,075,983 ops/sec ±0.31% (95 runs sampled)
mpath.has("comments.name", obj) x 4,073,711 ops/sec ±0.54% (89 runs sampled)
mpath.has("foo[1mystring]", obj) x 2,056,865 ops/sec ±0.24% (94 runs sampled)
mpath.has("foo[1mystring].bar[1]", obj) x 1,752,298 ops/sec ±0.36% (94 runs sampled)
mpath.has("foo[1mystring][2]", obj) x 1,927,552 ops/sec ±0.54% (92 runs sampled)
mpath.has(["comments", "2", "comments","1","comments","0","val"], obj) x 8,401,156 ops/sec ±0.78% (92 runs sampled)
mpath.set("first", obj, val) x 19,417,628 ops/sec ±1.04% (89 runs sampled)
mpath.set("first.second", obj, val) x 4,044,316 ops/sec ±0.28% (94 runs sampled)
mpath.set("first.second.third.1.name", obj, val) x 2,984,935 ops/sec ±0.28% (95 runs sampled)
mpath.set("comments", obj, val) x 13,202,177 ops/sec ±0.93% (89 runs sampled)
mpath.set("comments.1", obj, val) x 3,630,334 ops/sec ±0.69% (89 runs sampled)
mpath.set("comments.2.name", obj, val) x 3,394,605 ops/sec ±0.48% (92 runs sampled)
mpath.set("comments.2.comments.1.comments.0.val", obj, val) x 2,451,853 ops/sec ±0.30% (94 runs sampled)
mpath.set("comments.name", obj, val) x 3,142,130 ops/sec ±0.73% (93 runs sampled)
mpath.set("arr.test", [[42]], obj)) x 3,412,588 ops/sec ±0.27% (95 runs sampled)
To get some reliable benchmarks, replace in unset return delete cur[part]; with return true.
mpath.unset("first", obj) x 19,953,208 ops/sec ±0.69% (92 runs sampled)
mpath.unset("first.second", obj) x 5,080,429 ops/sec ±0.40% (93 runs sampled)
mpath.unset("first.second.third.1.name", obj) x 5,091,129 ops/sec ±0.42% (93 runs sampled)
mpath.unset("comments", obj) x 19,640,944 ops/sec ±0.75% (91 runs sampled)
mpath.unset("comments.1", obj) x 5,065,029 ops/sec ±0.40% (90 runs sampled)
mpath.unset("comments.2.name", obj) x 5,078,789 ops/sec ±0.44% (91 runs sampled)
mpath.unset("comments.2.comments.1.comments.0.val", obj) x 4,935,632 ops/sec ±0.34% (95 runs sampled)
mpath.unset("comments.name", obj) x 5,044,646 ops/sec ±0.37% (93 runs sampled)
```

Uzlopak commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,21 @@
+
+      - run: npm i
+      - run: npm run test
+      - run: npm run lint
```
removed linting

vkarpov15 commented on lib/index.js:
```diff
@@ -191,58 +241,48 @@ exports.set = function(path, val, o, special, map, _copying) {
 
-    if (Array.isArray(obj) && !/^\d+$/.test(part)) {
-      var paths = parts.slice(i);
+    if (Array.isArray(obj) && !numbers.has(part)) {
```
Unfortunately I don't think this will work, because what happens if you try to set `arr.11`? `numbers` only contains digits.

Uzlopak replied:
added unit test and fixed it

AbdelrahmanHafez commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,20 @@
+on: [push, pull_request]
+jobs:
+  test:
+    name: Node ${{ matrix.node }}, Mongo ${{ matrix.mongodb-version }}, on ${{ matrix.os }}
```
`Mongo ${{ matrix.mongodb-version }}`
I don't think mpath needs mongodb at all.

Uzlopak replied:
Good Catch, Copy Paste error

AbdelrahmanHafez commented on History.md:
```diff
@@ -1,3 +1,6 @@
+0.9.0 / TBA
+ * perf: Improve overall performance
```
Would help if we can quantify the performance improvement here.
You've done a great job and I'd like people to know that.

"Improve overall performance, up to 10x in some functionalities." might sell better.

Uzlopak replied:
Will have a thought on this.

vkarpov15 commented on lib/stringToParts.js:
```diff
@@ -1,46 +1,62 @@
     // square brackets contains something other than a number.
-    if (state === 'IN_SQUARE_BRACKETS' && !/\d/.test(str[i]) && str[i] !== ']') {
-      state = 'DEFAULT';
+    if (state === 1 && char !== ']' && !numbers.has(char)) {
```
Is using 1 instead of 'IN_SQUARE_BRACKETS' noticably faster? I'm not sure the readability tradeoff is worth the extra perf.

Uzlopak replied:
using now constants instead of magic numbers

vkarpov15 commented:
Re: https://github.com/aheckmann/mpath/pull/14#issuecomment-1016242711, "I think it is better to have a stable performance, than have a very fast degrading performance", I agree, predictable performance is better for this lib.

vkarpov15 reviewed (commented):
Thanks for the very detailed work! Overall looks good, and I appreciate the work. Just had a couple of comments/questions.

Uzlopak commented:
@vkarpov15 
@AbdelrahmanHafez 

I fixed your remarks. I added the ability to set an additional tokenizer cache. In my benchmark I use a simple Map, but you are able to use e.g. lru-cache. 

On node 14

```
> node ./benchmarks/bench.js

stringToParts("first") x 69,526,920 ops/sec ±0.36% (95 runs sampled)
stringToParts("first.second") x 6,540,082 ops/sec ±0.30% (94 runs sampled)
stringToParts("first.second.third.1.name") x 6,467,865 ops/sec ±0.56% (92 runs sampled)
stringToParts("comments") x 67,906,617 ops/sec ±0.88% (94 runs sampled)
stringToParts("comments.1") x 6,465,457 ops/sec ±0.59% (92 runs sampled)
stringToParts("comments.2.name") x 6,509,926 ops/sec ±0.44% (88 runs sampled)
stringToParts("comments.2.comments.1.comments.0.val") x 6,376,303 ops/sec ±0.67% (93 runs sampled)
stringToParts("comments.name") x 6,515,347 ops/sec ±0.28% (93 runs sampled)
stringToParts("foo[1mystring]") x 4,563,333 ops/sec ±0.69% (93 runs sampled)
stringToParts("foo[1mystring].bar[1]") x 3,248,934 ops/sec ±0.61% (88 runs sampled)
stringToParts("foo[1mystring][2]") x 3,983,136 ops/sec ±0.72% (92 runs sampled)
mpath.get("first", obj) x 33,572,570 ops/sec ±0.60% (93 runs sampled)
mpath.get("first.second", obj) x 4,717,555 ops/sec ±0.28% (92 runs sampled)
mpath.get("first.second.third.1.name", obj) x 3,434,640 ops/sec ±0.53% (91 runs sampled)
mpath.get("comments", obj) x 17,803,279 ops/sec ±0.69% (95 runs sampled)
mpath.get("comments.1", obj) x 4,074,737 ops/sec ±0.26% (95 runs sampled)
mpath.get("comments.2.name", obj) x 3,625,081 ops/sec ±0.43% (91 runs sampled)
mpath.get("comments.2.comments.1.comments.0.val", obj) x 2,682,924 ops/sec ±0.29% (96 runs sampled)
mpath.get("comments.name", obj) x 1,631,523 ops/sec ±0.29% (93 runs sampled)
mpath.has("first", obj) x 28,257,533 ops/sec ±1.28% (87 runs sampled)
mpath.has("first.second", obj) x 4,240,977 ops/sec ±0.72% (91 runs sampled)
mpath.has("first.second.third.1.name", obj) x 3,426,092 ops/sec ±0.43% (93 runs sampled)
mpath.has("comments", obj) x 17,531,812 ops/sec ±0.91% (89 runs sampled)
mpath.has("comments.1", obj) x 4,181,889 ops/sec ±0.56% (91 runs sampled)
mpath.has("comments.2.name", obj) x 3,759,199 ops/sec ±0.35% (93 runs sampled)
mpath.has("comments.2.comments.1.comments.0.val", obj) x 3,025,369 ops/sec ±0.45% (94 runs sampled)
mpath.has("comments.name", obj) x 3,937,535 ops/sec ±0.38% (92 runs sampled)
mpath.has("foo[1mystring]", obj) x 1,974,290 ops/sec ±0.49% (90 runs sampled)
mpath.has("foo[1mystring].bar[1]", obj) x 1,711,743 ops/sec ±0.41% (91 runs sampled)
mpath.has("foo[1mystring][2]", obj) x 1,900,625 ops/sec ±0.41% (91 runs sampled)
mpath.has(["comments", "2", "comments","1","comments","0","val"], obj) x 8,381,485 ops/sec ±0.67% (91 runs sampled)
mpath.set("first", obj, val) x 20,013,360 ops/sec ±0.56% (93 runs sampled)
mpath.set("first.second", obj, val) x 4,030,378 ops/sec ±0.78% (90 runs sampled)
mpath.set("first.second.third.1.name", obj, val) x 2,665,468 ops/sec ±0.41% (92 runs sampled)
mpath.set("comments", obj, val) x 13,445,133 ops/sec ±1.27% (89 runs sampled)
mpath.set("comments.1", obj, val) x 3,177,031 ops/sec ±0.97% (88 runs sampled)
mpath.set("comments.2.name", obj, val) x 3,083,730 ops/sec ±0.48% (94 runs sampled)
mpath.set("comments.2.comments.1.comments.0.val", obj, val) x 2,203,725 ops/sec ±0.36% (94 runs sampled)
mpath.set("comments.name", obj, val) x 2,985,744 ops/sec ±0.40% (95 runs sampled)
mpath.set("arr.test", [[42]], obj)) x 3,188,097 ops/sec ±0.30% (93 runs sampled)
To get some reliable benchmarks, replace in unset return delete cur[part]; with return true.
mpath.unset("first", obj) x 17,863,091 ops/sec ±1.07% (84 runs sampled)
mpath.unset("first.second", obj) x 4,648,304 ops/sec ±0.64% (91 runs sampled)
mpath.unset("first.second.third.1.name", obj) x 4,606,684 ops/sec ±0.41% (93 runs sampled)
mpath.unset("comments", obj) x 19,293,496 ops/sec ±0.80% (91 runs sampled)
mpath.unset("comments.1", obj) x 5,006,040 ops/sec ±0.38% (95 runs sampled)
mpath.unset("comments.2.name", obj) x 5,031,569 ops/sec ±0.36% (92 runs sampled)
mpath.unset("comments.2.comments.1.comments.0.val", obj) x 5,028,283 ops/sec ±0.40% (92 runs sampled)
mpath.unset("comments.name", obj) x 5,163,328 ops/sec ±0.42% (91 runs sampled)
```

```
> node ./benchmarks/bench-with-cache.js

stringToParts("first") x 55,035,746 ops/sec ±0.29% (95 runs sampled)
stringToParts("first.second") x 124,344,121 ops/sec ±0.61% (96 runs sampled)
stringToParts("first.second.third.1.name") x 125,387,034 ops/sec ±0.93% (97 runs sampled)
stringToParts("comments") x 52,652,732 ops/sec ±0.39% (96 runs sampled)
stringToParts("comments.1") x 123,813,267 ops/sec ±0.66% (95 runs sampled)
stringToParts("comments.2.name") x 122,354,410 ops/sec ±0.31% (94 runs sampled)
stringToParts("comments.2.comments.1.comments.0.val") x 125,275,601 ops/sec ±0.56% (94 runs sampled)
stringToParts("comments.name") x 124,528,396 ops/sec ±0.98% (93 runs sampled)
stringToParts("foo[1mystring]") x 100,125,510 ops/sec ±6.57% (78 runs sampled)
stringToParts("foo[1mystring].bar[1]") x 67,375,712 ops/sec ±2.31% (85 runs sampled)
stringToParts("foo[1mystring][2]") x 122,988,971 ops/sec ±0.89% (95 runs sampled)
mpath.get("first", obj) x 26,438,262 ops/sec ±1.09% (93 runs sampled)
mpath.get("first.second", obj) x 14,638,445 ops/sec ±1.27% (93 runs sampled)
mpath.get("first.second.third.1.name", obj) x 8,258,785 ops/sec ±0.97% (88 runs sampled)
mpath.get("comments", obj) x 15,823,542 ops/sec ±0.78% (91 runs sampled)
mpath.get("comments.1", obj) x 10,922,029 ops/sec ±0.63% (93 runs sampled)
mpath.get("comments.2.name", obj) x 9,441,785 ops/sec ±2.23% (82 runs sampled)
mpath.get("comments.2.comments.1.comments.0.val", obj) x 5,125,201 ops/sec ±1.01% (86 runs sampled)
mpath.get("comments.name", obj) x 2,352,914 ops/sec ±0.38% (95 runs sampled)
mpath.has("first", obj) x 23,823,239 ops/sec ±0.99% (90 runs sampled)
mpath.has("first.second", obj) x 14,094,837 ops/sec ±1.40% (89 runs sampled)
mpath.has("first.second.third.1.name", obj) x 9,633,501 ops/sec ±1.61% (87 runs sampled)
mpath.has("comments", obj) x 15,897,412 ops/sec ±0.82% (91 runs sampled)
mpath.has("comments.1", obj) x 12,590,462 ops/sec ±1.09% (89 runs sampled)
mpath.has("comments.2.name", obj) x 13,234,734 ops/sec ±1.40% (83 runs sampled)
mpath.has("comments.2.comments.1.comments.0.val", obj) x 7,586,033 ops/sec ±0.51% (95 runs sampled)
mpath.has("comments.name", obj) x 10,863,456 ops/sec ±1.03% (92 runs sampled)
mpath.has("foo[1mystring]", obj) x 9,732,172 ops/sec ±1.02% (84 runs sampled)
mpath.has("foo[1mystring].bar[1]", obj) x 10,220,312 ops/sec ±0.73% (93 runs sampled)
mpath.has("foo[1mystring][2]", obj) x 10,172,951 ops/sec ±0.75% (92 runs sampled)
mpath.has(["comments", "2", "comments","1","comments","0","val"], obj) x 8,524,693 ops/sec ±0.61% (91 runs sampled)
mpath.set("first", obj, val) x 16,819,567 ops/sec ±1.01% (92 runs sampled)
mpath.set("first.second", obj, val) x 9,351,171 ops/sec ±0.57% (92 runs sampled)
mpath.set("first.second.third.1.name", obj, val) x 4,925,642 ops/sec ±0.38% (92 runs sampled)
mpath.set("comments", obj, val) x 11,523,814 ops/sec ±0.78% (88 runs sampled)
mpath.set("comments.1", obj, val) x 6,557,040 ops/sec ±1.18% (92 runs sampled)
mpath.set("comments.2.name", obj, val) x 6,460,976 ops/sec ±0.56% (93 runs sampled)
mpath.set("comments.2.comments.1.comments.0.val", obj, val) x 3,633,191 ops/sec ±0.28% (92 runs sampled)
mpath.set("comments.name", obj, val) x 6,115,856 ops/sec ±0.89% (89 runs sampled)
mpath.set("arr.test", [[42]], obj)) x 6,845,332 ops/sec ±0.65% (93 runs sampled)
To get some reliable benchmarks, replace in unset return delete cur[part]; with return true.
mpath.unset("first", obj) x 17,165,169 ops/sec ±0.75% (93 runs sampled)
mpath.unset("first.second", obj) x 18,900,890 ops/sec ±1.23% (90 runs sampled)
mpath.unset("first.second.third.1.name", obj) x 19,776,773 ops/sec ±1.29% (86 runs sampled)
mpath.unset("comments", obj) x 14,991,654 ops/sec ±1.05% (92 runs sampled)
mpath.unset("comments.1", obj) x 16,700,442 ops/sec ±1.40% (89 runs sampled)
mpath.unset("comments.2.name", obj) x 23,575,072 ops/sec ±1.33% (89 runs sampled)
mpath.unset("comments.2.comments.1.comments.0.val", obj) x 18,612,758 ops/sec ±1.03% (80 runs sampled)
mpath.unset("comments.name", obj) x 15,589,431 ops/sec ±0.58% (90 runs sampled)
```

In production you would of course use e.g. lru-cache

Uzlopak commented:
Ok... I think I am done. I dont know how I can improve the speed any further :).

I ensured that if you dont use caching, the performance wont get hurt. I added another benchmark for fast-lru. 

Here the latest benchmarks on node 14.

```
> node ./benchmarks/bench.js

tokenize("first") x 73,602,314 ops/sec ±0.35% (95 runs sampled)
tokenize("first.second") x 6,653,143 ops/sec ±0.61% (95 runs sampled)
tokenize("first.second.third.1.name") x 6,633,484 ops/sec ±0.32% (95 runs sampled)
tokenize("comments") x 71,939,005 ops/sec ±0.95% (95 runs sampled)
tokenize("comments.1") x 6,683,093 ops/sec ±0.35% (94 runs sampled)
tokenize("comments.2.name") x 6,642,379 ops/sec ±0.30% (94 runs sampled)
tokenize("comments.2.comments.1.comments.0.val") x 6,507,134 ops/sec ±0.32% (95 runs sampled)
tokenize("comments.name") x 6,643,137 ops/sec ±0.38% (91 runs sampled)
tokenize("foo[1mystring]") x 4,512,247 ops/sec ±0.66% (87 runs sampled)
tokenize("foo[1mystring].bar[1]") x 3,344,177 ops/sec ±0.60% (92 runs sampled)
tokenize("foo[1mystring][2]") x 4,047,169 ops/sec ±0.98% (92 runs sampled)
mpath.get("first", obj) x 40,151,591 ops/sec ±0.40% (97 runs sampled)
mpath.get("first.second", obj) x 4,417,092 ops/sec ±0.30% (94 runs sampled)
mpath.get("first.second.third.1.name", obj) x 3,137,493 ops/sec ±0.52% (91 runs sampled)
mpath.get("comments", obj) x 21,629,229 ops/sec ±2.37% (91 runs sampled)
mpath.get("comments.1", obj) x 3,642,275 ops/sec ±0.45% (91 runs sampled)
mpath.get("comments.2.name", obj) x 3,346,850 ops/sec ±0.59% (89 runs sampled)
mpath.get("comments.2.comments.1.comments.0.val", obj) x 2,471,729 ops/sec ±1.13% (92 runs sampled)
mpath.get("comments.name", obj) x 1,353,844 ops/sec ±1.09% (85 runs sampled)
mpath.has("first", obj) x 38,068,509 ops/sec ±4.69% (76 runs sampled)
mpath.has("first.second", obj) x 4,301,352 ops/sec ±0.43% (90 runs sampled)
mpath.has("first.second.third.1.name", obj) x 3,435,863 ops/sec ±0.42% (94 runs sampled)
mpath.has("comments", obj) x 20,141,263 ops/sec ±2.11% (79 runs sampled)
mpath.has("comments.1", obj) x 4,183,104 ops/sec ±0.37% (91 runs sampled)
mpath.has("comments.2.name", obj) x 3,814,918 ops/sec ±0.38% (93 runs sampled)
mpath.has("comments.2.comments.1.comments.0.val", obj) x 3,071,276 ops/sec ±0.35% (94 runs sampled)
mpath.has("comments.name", obj) x 4,015,664 ops/sec ±0.37% (94 runs sampled)
mpath.has("foo[1mystring]", obj) x 2,043,284 ops/sec ±0.59% (94 runs sampled)
mpath.has("foo[1mystring].bar[1]", obj) x 1,755,779 ops/sec ±0.42% (89 runs sampled)
mpath.has("foo[1mystring][2]", obj) x 1,910,760 ops/sec ±0.50% (96 runs sampled)
mpath.has(["comments", "2", "comments","1","comments","0","val"], obj) x 8,552,324 ops/sec ±1.58% (89 runs sampled)
mpath.set("first", obj, val) x 20,171,790 ops/sec ±0.77% (92 runs sampled)
mpath.set("first.second", obj, val) x 4,011,799 ops/sec ±0.38% (94 runs sampled)
mpath.set("first.second.third.1.name", obj, val) x 2,705,552 ops/sec ±0.75% (90 runs sampled)
mpath.set("comments", obj, val) x 13,949,810 ops/sec ±1.07% (93 runs sampled)
mpath.set("comments.1", obj, val) x 3,369,428 ops/sec ±0.52% (91 runs sampled)
mpath.set("comments.2.name", obj, val) x 3,066,993 ops/sec ±0.39% (92 runs sampled)
mpath.set("comments.2.comments.1.comments.0.val", obj, val) x 2,133,967 ops/sec ±0.25% (95 runs sampled)
mpath.set("comments.name", obj, val) x 3,045,378 ops/sec ±0.31% (92 runs sampled)
mpath.set("arr.test", [[42]], obj)) x 3,188,002 ops/sec ±0.27% (95 runs sampled)
To get some reliable benchmarks, replace in unset return delete cur[part]; with return true.
mpath.unset("first", obj) x 20,145,459 ops/sec ±1.21% (89 runs sampled)
mpath.unset("first.second", obj) x 5,141,114 ops/sec ±0.40% (93 runs sampled)
mpath.unset("first.second.third.1.name", obj) x 5,058,658 ops/sec ±0.31% (96 runs sampled)
mpath.unset("comments", obj) x 20,190,378 ops/sec ±1.03% (84 runs sampled)
mpath.unset("comments.1", obj) x 5,037,352 ops/sec ±0.55% (94 runs sampled)
mpath.unset("comments.2.name", obj) x 5,060,874 ops/sec ±0.41% (94 runs sampled)
mpath.unset("comments.2.comments.1.comments.0.val", obj) x 5,083,899 ops/sec ±0.54% (94 runs sampled)
mpath.unset("comments.name", obj) x 5,097,223 ops/sec ±0.46% (91 runs sampled)
```

```
> node ./benchmarks/bench-map.js

tokenize("first") x 119,427,554 ops/sec ±0.92% (92 runs sampled)
tokenize("first.second") x 118,660,542 ops/sec ±0.81% (92 runs sampled)
tokenize("first.second.third.1.name") x 117,976,287 ops/sec ±0.91% (91 runs sampled)
tokenize("comments") x 119,042,523 ops/sec ±0.70% (92 runs sampled)
tokenize("comments.1") x 118,511,372 ops/sec ±0.80% (90 runs sampled)
tokenize("comments.2.name") x 117,701,226 ops/sec ±0.90% (93 runs sampled)
tokenize("comments.2.comments.1.comments.0.val") x 115,852,937 ops/sec ±1.40% (91 runs sampled)
tokenize("comments.name") x 116,984,552 ops/sec ±1.12% (93 runs sampled)
tokenize("foo[1mystring]") x 118,267,634 ops/sec ±1.38% (94 runs sampled)
tokenize("foo[1mystring].bar[1]") x 77,620,970 ops/sec ±7.28% (66 runs sampled)
tokenize("foo[1mystring][2]") x 114,087,019 ops/sec ±0.65% (95 runs sampled)
mpath.get("first", obj) x 67,350,325 ops/sec ±1.71% (90 runs sampled)
mpath.get("first.second", obj) x 19,136,542 ops/sec ±0.53% (90 runs sampled)
mpath.get("first.second.third.1.name", obj) x 8,054,298 ops/sec ±1.14% (86 runs sampled)
mpath.get("comments", obj) x 20,022,972 ops/sec ±2.22% (81 runs sampled)
mpath.get("comments.1", obj) x 13,664,586 ops/sec ±1.50% (86 runs sampled)
mpath.get("comments.2.name", obj) x 10,130,770 ops/sec ±1.54% (83 runs sampled)
mpath.get("comments.2.comments.1.comments.0.val", obj) x 4,911,963 ops/sec ±0.82% (90 runs sampled)
mpath.get("comments.name", obj) x 2,499,963 ops/sec ±0.39% (95 runs sampled)
mpath.has("first", obj) x 51,661,206 ops/sec ±1.99% (85 runs sampled)
mpath.has("first.second", obj) x 18,008,804 ops/sec ±2.85% (82 runs sampled)
mpath.has("first.second.third.1.name", obj) x 10,305,092 ops/sec ±1.18% (89 runs sampled)
mpath.has("comments", obj) x 22,988,539 ops/sec ±2.41% (85 runs sampled)
mpath.has("comments.1", obj) x 18,740,067 ops/sec ±2.15% (79 runs sampled)
mpath.has("comments.2.name", obj) x 14,143,804 ops/sec ±1.60% (89 runs sampled)
mpath.has("comments.2.comments.1.comments.0.val", obj) x 7,885,299 ops/sec ±0.93% (91 runs sampled)
mpath.has("comments.name", obj) x 14,976,906 ops/sec ±1.92% (78 runs sampled)
mpath.has("foo[1mystring]", obj) x 8,562,702 ops/sec ±0.48% (94 runs sampled)
mpath.has("foo[1mystring].bar[1]", obj) x 8,852,109 ops/sec ±0.66% (93 runs sampled)
mpath.has("foo[1mystring][2]", obj) x 8,979,580 ops/sec ±0.54% (85 runs sampled)
mpath.has(["comments", "2", "comments","1","comments","0","val"], obj) x 8,787,499 ops/sec ±0.66% (90 runs sampled)
mpath.set("first", obj, val) x 25,249,135 ops/sec ±1.09% (88 runs sampled)
mpath.set("first.second", obj, val) x 12,579,603 ops/sec ±0.70% (92 runs sampled)
mpath.set("first.second.third.1.name", obj, val) x 5,389,175 ops/sec ±0.57% (94 runs sampled)
mpath.set("comments", obj, val) x 14,762,096 ops/sec ±1.32% (92 runs sampled)
mpath.set("comments.1", obj, val) x 9,261,277 ops/sec ±0.57% (91 runs sampled)
mpath.set("comments.2.name", obj, val) x 7,903,874 ops/sec ±0.75% (92 runs sampled)
mpath.set("comments.2.comments.1.comments.0.val", obj, val) x 4,125,593 ops/sec ±0.39% (94 runs sampled)
mpath.set("comments.name", obj, val) x 7,836,323 ops/sec ±0.87% (90 runs sampled)
mpath.set("arr.test", [[42]], obj)) x 6,715,982 ops/sec ±0.45% (94 runs sampled)
To get some reliable benchmarks, replace in unset return delete cur[part]; with return true.
mpath.unset("first", obj) x 30,230,903 ops/sec ±1.99% (89 runs sampled)
mpath.unset("first.second", obj) x 25,048,250 ops/sec ±2.46% (86 runs sampled)
mpath.unset("first.second.third.1.name", obj) x 30,069,523 ops/sec ±2.09% (86 runs sampled)
mpath.unset("comments", obj) x 24,948,306 ops/sec ±2.64% (87 runs sampled)
mpath.unset("comments.1", obj) x 26,954,343 ops/sec ±2.71% (85 runs sampled)
mpath.unset("comments.2.name", obj) x 30,262,407 ops/sec ±2.56% (88 runs sampled)
mpath.unset("comments.2.comments.1.comments.0.val", obj) x 30,309,162 ops/sec ±2.14% (90 runs sampled)
mpath.unset("comments.name", obj) x 30,403,412 ops/sec ±1.67% (87 runs sampled)
```

```
> node ./benchmarks/bench-lru.js

tokenize("first") x 78,565,481 ops/sec ±0.19% (96 runs sampled)
tokenize("first.second") x 76,301,576 ops/sec ±0.96% (89 runs sampled)
tokenize("first.second.third.1.name") x 76,752,938 ops/sec ±0.63% (93 runs sampled)
tokenize("comments") x 76,718,757 ops/sec ±0.61% (90 runs sampled)
tokenize("comments.1") x 77,216,999 ops/sec ±0.52% (92 runs sampled)
tokenize("comments.2.name") x 78,738,757 ops/sec ±0.30% (91 runs sampled)
tokenize("comments.2.comments.1.comments.0.val") x 79,546,661 ops/sec ±0.55% (92 runs sampled)
tokenize("comments.name") x 78,946,878 ops/sec ±0.84% (92 runs sampled)
tokenize("foo[1mystring]") x 79,946,830 ops/sec ±0.47% (94 runs sampled)
tokenize("foo[1mystring].bar[1]") x 79,432,349 ops/sec ±0.52% (95 runs sampled)
tokenize("foo[1mystring][2]") x 68,379,280 ops/sec ±5.20% (83 runs sampled)
mpath.get("first", obj) x 39,745,173 ops/sec ±1.47% (89 runs sampled)
mpath.get("first.second", obj) x 18,572,919 ops/sec ±1.13% (93 runs sampled)
mpath.get("first.second.third.1.name", obj) x 8,525,283 ops/sec ±0.70% (92 runs sampled)
mpath.get("comments", obj) x 22,134,515 ops/sec ±1.86% (82 runs sampled)
mpath.get("comments.1", obj) x 11,333,987 ops/sec ±0.50% (91 runs sampled)
mpath.get("comments.2.name", obj) x 9,603,373 ops/sec ±0.78% (90 runs sampled)
mpath.get("comments.2.comments.1.comments.0.val", obj) x 5,550,875 ops/sec ±0.75% (95 runs sampled)
mpath.get("comments.name", obj) x 2,411,148 ops/sec ±0.34% (97 runs sampled)
mpath.has("first", obj) x 43,554,156 ops/sec ±2.22% (87 runs sampled)
mpath.has("first.second", obj) x 15,476,908 ops/sec ±2.16% (84 runs sampled)
mpath.has("first.second.third.1.name", obj) x 10,903,294 ops/sec ±0.94% (89 runs sampled)
mpath.has("comments", obj) x 22,291,814 ops/sec ±2.82% (83 runs sampled)
mpath.has("comments.1", obj) x 17,515,340 ops/sec ±2.12% (85 runs sampled)
mpath.has("comments.2.name", obj) x 13,063,961 ops/sec ±2.19% (83 runs sampled)
mpath.has("comments.2.comments.1.comments.0.val", obj) x 7,772,087 ops/sec ±0.93% (91 runs sampled)
mpath.has("comments.name", obj) x 12,714,557 ops/sec ±1.28% (90 runs sampled)
mpath.has("foo[1mystring]", obj) x 10,634,882 ops/sec ±0.96% (85 runs sampled)
mpath.has("foo[1mystring].bar[1]", obj) x 10,466,770 ops/sec ±0.88% (93 runs sampled)
mpath.has("foo[1mystring][2]", obj) x 10,526,990 ops/sec ±0.83% (91 runs sampled)
mpath.has(["comments", "2", "comments","1","comments","0","val"], obj) x 8,785,434 ops/sec ±0.73% (93 runs sampled)
mpath.set("first", obj, val) x 22,569,056 ops/sec ±0.99% (91 runs sampled)
mpath.set("first.second", obj, val) x 9,681,512 ops/sec ±1.89% (86 runs sampled)
mpath.set("first.second.third.1.name", obj, val) x 5,420,669 ops/sec ±0.40% (93 runs sampled)
mpath.set("comments", obj, val) x 15,424,779 ops/sec ±0.49% (94 runs sampled)
mpath.set("comments.1", obj, val) x 7,594,027 ops/sec ±0.76% (91 runs sampled)
mpath.set("comments.2.name", obj, val) x 5,833,113 ops/sec ±0.34% (95 runs sampled)
mpath.set("comments.2.comments.1.comments.0.val", obj, val) x 3,074,046 ops/sec ±0.27% (93 runs sampled)
mpath.set("comments.name", obj, val) x 6,773,152 ops/sec ±0.65% (94 runs sampled)
mpath.set("arr.test", [[42]], obj)) x 6,869,394 ops/sec ±0.54% (94 runs sampled)
To get some reliable benchmarks, replace in unset return delete cur[part]; with return true.
mpath.unset("first", obj) x 27,365,755 ops/sec ±2.01% (92 runs sampled)
mpath.unset("first.second", obj) x 22,100,363 ops/sec ±3.08% (84 runs sampled)
mpath.unset("first.second.third.1.name", obj) x 27,436,755 ops/sec ±1.89% (88 runs sampled)
mpath.unset("comments", obj) x 22,448,781 ops/sec ±2.74% (85 runs sampled)
mpath.unset("comments.1", obj) x 25,760,376 ops/sec ±1.86% (89 runs sampled)
mpath.unset("comments.2.name", obj) x 22,357,883 ops/sec ±2.90% (83 runs sampled)
mpath.unset("comments.2.comments.1.comments.0.val", obj) x 25,384,935 ops/sec ±2.42% (87 runs sampled)
mpath.unset("comments.name", obj) x 21,294,295 ops/sec ±2.72% (87 runs sampled)
```

```
> node bench-fast-lru.js 
tokenize("first") x 80,119,006 ops/sec ±0.42% (93 runs sampled)
tokenize("first.second") x 78,265,237 ops/sec ±0.31% (89 runs sampled)
tokenize("first.second.third.1.name") x 77,256,486 ops/sec ±1.18% (95 runs sampled)
tokenize("comments") x 78,519,284 ops/sec ±0.26% (94 runs sampled)
tokenize("comments.1") x 78,244,371 ops/sec ±0.57% (95 runs sampled)
tokenize("comments.2.name") x 78,652,327 ops/sec ±0.22% (94 runs sampled)
tokenize("comments.2.comments.1.comments.0.val") x 78,405,391 ops/sec ±0.34% (90 runs sampled)
tokenize("comments.name") x 78,706,567 ops/sec ±0.15% (95 runs sampled)
tokenize("foo[1mystring]") x 73,550,946 ops/sec ±2.98% (89 runs sampled)
tokenize("foo[1mystring].bar[1]") x 49,859,259 ops/sec ±1.75% (85 runs sampled)
tokenize("foo[1mystring][2]") x 75,926,118 ops/sec ±0.41% (95 runs sampled)
mpath.get("first", obj) x 59,887,267 ops/sec ±0.25% (97 runs sampled)
mpath.get("first.second", obj) x 18,115,810 ops/sec ±0.31% (95 runs sampled)
mpath.get("first.second.third.1.name", obj) x 7,336,169 ops/sec ±1.36% (83 runs sampled)
mpath.get("comments", obj) x 22,251,326 ops/sec ±1.42% (92 runs sampled)
mpath.get("comments.1", obj) x 12,460,248 ops/sec ±1.39% (86 runs sampled)
mpath.get("comments.2.name", obj) x 9,621,482 ops/sec ±2.64% (81 runs sampled)
mpath.get("comments.2.comments.1.comments.0.val", obj) x 4,876,255 ops/sec ±1.66% (85 runs sampled)
mpath.get("comments.name", obj) x 2,338,906 ops/sec ±0.47% (92 runs sampled)
mpath.has("first", obj) x 45,249,502 ops/sec ±1.41% (88 runs sampled)
mpath.has("first.second", obj) x 17,095,561 ops/sec ±3.19% (83 runs sampled)
mpath.has("first.second.third.1.name", obj) x 9,785,630 ops/sec ±1.37% (88 runs sampled)
mpath.has("comments", obj) x 26,945,777 ops/sec ±2.61% (78 runs sampled)
mpath.has("comments.1", obj) x 16,020,595 ops/sec ±2.12% (88 runs sampled)
mpath.has("comments.2.name", obj) x 12,986,558 ops/sec ±1.85% (88 runs sampled)
mpath.has("comments.2.comments.1.comments.0.val", obj) x 7,925,644 ops/sec ±0.76% (91 runs sampled)
mpath.has("comments.name", obj) x 14,081,286 ops/sec ±1.77% (84 runs sampled)
mpath.has("foo[1mystring]", obj) x 10,511,316 ops/sec ±0.88% (90 runs sampled)
mpath.has("foo[1mystring].bar[1]", obj) x 10,605,603 ops/sec ±0.95% (86 runs sampled)
mpath.has("foo[1mystring][2]", obj) x 10,472,399 ops/sec ±0.85% (92 runs sampled)
mpath.has(["comments", "2", "comments","1","comments","0","val"], obj) x 8,754,251 ops/sec ±0.68% (92 runs sampled)
mpath.set("first", obj, val) x 24,368,354 ops/sec ±1.09% (77 runs sampled)
mpath.set("first.second", obj, val) x 12,281,550 ops/sec ±0.82% (90 runs sampled)
mpath.set("first.second.third.1.name", obj, val) x 5,561,399 ops/sec ±0.60% (94 runs sampled)
mpath.set("comments", obj, val) x 16,616,562 ops/sec ±1.09% (91 runs sampled)
mpath.set("comments.1", obj, val) x 8,473,119 ops/sec ±0.64% (93 runs sampled)
mpath.set("comments.2.name", obj, val) x 6,498,360 ops/sec ±0.52% (94 runs sampled)
mpath.set("comments.2.comments.1.comments.0.val", obj, val) x 3,823,884 ops/sec ±0.46% (91 runs sampled)
mpath.set("comments.name", obj, val) x 6,691,750 ops/sec ±0.89% (93 runs sampled)
mpath.set("arr.test", [[42]], obj)) x 6,932,162 ops/sec ±0.34% (94 runs sampled)
To get some reliable benchmarks, replace in unset return delete cur[part]; with return true.
mpath.unset("first", obj) x 28,115,245 ops/sec ±1.91% (85 runs sampled)
mpath.unset("first.second", obj) x 24,723,464 ops/sec ±1.17% (91 runs sampled)
mpath.unset("first.second.third.1.name", obj) x 24,373,236 ops/sec ±2.67% (84 runs sampled)
mpath.unset("comments", obj) x 27,049,606 ops/sec ±1.96% (86 runs sampled)
mpath.unset("comments.1", obj) x 23,514,034 ops/sec ±2.60% (87 runs sampled)
mpath.unset("comments.2.name", obj) x 23,906,961 ops/sec ±2.74% (83 runs sampled)
mpath.unset("comments.2.comments.1.comments.0.val", obj) x 26,186,249 ops/sec ±2.65% (88 runs sampled)
mpath.unset("comments.name", obj) x 27,175,413 ops/sec ±1.92% (89 runs sampled)
```

Uzlopak commented on lib/tokenize.js:
```diff
@@ -0,0 +1,87 @@
+}
+
+function tokenizeCached(str) {
+  const result = cache.get(str);
```
the intermediate variable can be avoided by replacing the function body with


  ```
return (
    cache.get('str') || 
    (cache.set(str, tokenize(str)) && false) ||
    cache.get('str')
  )
```

But it seems like the overall performance gets hurt, as it seems that v8 can not optimize the function properly. Performance hit of about 40 %.

Uzlopak commented:
@kibertoad
@mcollina

*wroom* *wroom* 

Any idea how to improve the perf further?

mcollina commented on lib/tokenize.js:
```diff
@@ -0,0 +1,87 @@
+
+exports.tokenize = tokenize;
+
+exports.setTokenizeCache = function(parmCache) {
```
Avoid this and use an LRU-cache from start, the one from mnemonist is great.

Uzlopak replied:
@mcollina Yes! mnemonist is truelly faster.

Should I use the new Function Syntax only for the tokenize method? I use some one time initialized sets and the cache, but it seems that I have to somehow pass them into the method? Do you have some example, where i could look how to it properly?
Thx.

mcollina replied:
Check out the fast-json-stringify module. I'd cache the full thing.

Uzlopak replied:
@mcollina 
I unfortunately don't know how I could apply your suggested change. I know you are a busy man, but could you help me out by giving me the code solution please? I am eager to learn as I want to improve more packages over the time, so it would be not a waste of your time, imho.

Please.

mcollina reviewed (commented):
I would improve perf by using new Function to compile functions to do this for me. You would see quite a massive performance boost as those would become as expensive as a normal property lookup.

kibertoad commented:
> I would improve perf by using new Function to compile functions to do this for me. You would see quite a massive performance boost as those would become as expensive as a normal property lookup.

@mcollina Could you please clarify which particular part of the code would benefit from this change?

mcollina commented:
The lookup of the property is currently done iterating an array. However it could be generated as a function, cached and called.

Uzlopak commented:
@mcollina 

I played a little bit around and have to say that this is a great performance booster. Thank you for this great idea. 

@vkarpov15 
Despite the great idea of mcollina: would you please consider to review and merge it? This PR already means a great performance boost for the package.
I could implement the suggestion of mcollina in the next iteration, version 1.0.0 :). 
We could actually already use the optimized tokenize method in various places of mongoose to replace the path.split(".") and gain potentially performance.

Uzlopak commented:
Hi @vkarpov15 

What is the goal for the future?

Should I refactor the package completely?

vkarpov15 commented on lib/tokenize.js:
```diff
@@ -0,0 +1,65 @@
+const IN_SQUARE_BRACKETS = Symbol('IN_SQUARE_BRACKETS');
+const IMMEDIATELY_AFTER_SQUARE_BRACKETS = Symbol('IMMEDIATELY_AFTER_SQUARE_BRACKETS');
+
+exports.tokenize = function tokenize(str) {
```
Can we keep the name as `stringToParts()`? Admittedly `tokenize()` is probably better. But I introduced some code that exports `stringToParts()`, so changing the name would be backwards breaking :(

Uzlopak replied:
Is an alias enough?

Uzlopak replied:
Done

vkarpov15 commented:
@Uzlopak sorry for missing the convo on this PR. The idea of precompiling a function is an interesting one, but the problem is that this package needs to support adhoc lookups. I don't want to have to do something Ramda-like where I need to create a function to access a given property, and then store that function. We potentially could do that with Mongoose by caching functions for every schema path, but I'm not sure the complexity is worth it.

Either way, I like this PR as is. Minus the one issue with `tokenize()`.

Uzlopak commented:
@vkarpov15 
Done

Uzlopak commented:
@vkarpov15 

I dont know ramda. But the the performance of mpath depends on two parts: The tokenization and the property access. While tokenization is an easy gain with caching, you can significantly improve the performance by avoiding tokenizing repeatedly. But the bigger performance gain can be achieved by actually building functoins on the fly. I tried it and if done simple cases like just accessing non Array attributes the performance is indeed like matteo writes crazy fast.

See here:
https://github.com/Uzlopak/mpath2/blob/master/lib/get.js

```
uzlopak@Battlestation:~/workspace/mpath2$ node benchmark/get.js 
mpath.get("first", obj) x 93,118,411 ops/sec ±0.33% (94 runs sampled)
mpath.get("first.second", obj) x 62,086,489 ops/sec ±0.50% (92 runs sampled)
mpath.get("first.second.third", obj) x 61,563,902 ops/sec ±0.50% (93 runs sampled)
mpath.get("first", obj) x 62,156,817 ops/sec ±1.06% (97 runs sampled)
```

So there is no real difference between accessing a one parter and a three parter. 

Issues arise when trying to write super generic code to handle arrays and combining them. I guess i would need still some time to figure it out. 

But still, matteos approach makes it crazy fast. But the trade-off is of course memory. And by reading your comments on mongoose I guess, a memory trade-off is not desired. 

But if you ever want to do the trade off, let me know.
